Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.cs

```diff
@@ -207,40 +226,52 @@ private void StopShooting(EntityUid uid, GunComponent gun)
     /// <summary>
     /// Attempts to shoot at the target coordinates. Resets the shot counter after every shot.
     /// </summary>
-    public void AttemptShoot(EntityUid user, EntityUid gunUid, GunComponent gun, EntityCoordinates toCoordinates)
+    public void AttemptShoot(EntityUid? user, EntityUid gunUid, GunComponent gun, EntityCoordinates? toCoordinates = null)
     {
-        gun.ShootCoordinates = toCoordinates;
+        if (toCoordinates == null)
+            gun.ShootCoordinates = new EntityCoordinates(gunUid, new Vector2(0, -1)); //Forward vector
+        else
+            gun.ShootCoordinates = toCoordinates;
+
         AttemptShoot(user, gunUid, gun);
         gun.ShotCounter = 0;
     }
 
-    private void AttemptShoot(EntityUid user, EntityUid gunUid, GunComponent gun)
+    private void AttemptShoot(EntityUid? user, EntityUid gunUid, GunComponent gun)
     {
-        if (gun.FireRate <= 0f ||
-            !_actionBlockerSystem.CanUseHeldEntity(user))
+        if (!gun.CanShootWithoutUser && user == null)
+            return;
+
+        if (gun.FireRate <= 0f)
             return;
 
+        if (user != null)
+        {
+            if (!_actionBlockerSystem.CanUseHeldEntity(user.Value))
+                return;
+
+            // check if anything wants to prevent shooting
+            var prevention = new ShotAttemptedEvent
+            {
+                User = user.Value,
+                Used = gunUid
+            };
+            RaiseLocalEvent(gunUid, ref prevention);
+            if (prevention.Cancelled)
+                return;
+
+            RaiseLocalEvent(user.Value, ref prevention);
+            if (prevention.Cancelled)
+                return;
+        }
+
         var toCoordinates = gun.ShootCoordinates;
 
         if (toCoordinates == null)
             return;
 
         var curTime = Timing.CurTime;
```

Comment:
This null check does not make sense anymore if you're going to pass entitycoordinates in, just use the gun shootcoordinates instead of adding another separate codepath.